# Cuaderno Jupyter: Parcial de Fundamentos de Programación


## Aplicado a Teoría Elemental de Números y Ciencia de Datos




# Sección 1: Condicionales `if`, `elif`, `else` (20 pts)

### Ejercicio 1.1 — Clasificador de números por suma de divisores (10 pts)

Escriba un programa que reciba un entero positivo `n` y determine si es:
- **Perfecto**: la suma de sus divisores propios es igual a `n`.
- **Abundante**: la suma de sus divisores propios es mayor que `n`.
- **Deficiente**: la suma de sus divisores propios es menor que `n`.

Además, si el número es **primo**, agregue `" y es primo"` al mensaje.

Pruebe con `n = 28`.

```python
# Celda 1.1
n = 28

# TU CÓDIGO AQUÍ
```



### Ejercicio 1.2 — Clasificador de notas con curva (10 pts)

Dado un puntaje `p` (0–100) y una curva `curva`, clasifique la nota final:
- `>= 90` → `"Excelente"`
- `>= 80` → `"Bueno"`
- `>= 70` → `"Aceptable"`
- `>= 60` → `"Insuficiente"`
- `< 60` → `"Reprobado"`

**Reglas especiales:**
- El puntaje final no puede exceder 100.
- Si el puntaje original es `< 60` pero con la curva alcanza `>= 60`, indique `"Aprobado por curva"`.

```python
# Celda 1.2
p = 55
curva = 8

# TU CÓDIGO AQUÍ
```



# Sección 2: Ciclos `for` y `while` (25 pts)

### Ejercicio 2.1 — Fibonacci + primalidad con `for` (12 pts)

Usando un ciclo `for`, genere los primeros `N = 20` términos de la sucesión de Fibonacci. Luego, determine cuáles de esos términos son **primos** e imprima la lista de Fibonacci primos.

⚠️ **No use `sorted()`**; si necesita ordenar algo, hágalo manualmente con ciclos.

```python
# Celda 2.1
N = 20

# TU CÓDIGO AQUÍ
```


In [5]:
f1 = 1
f2 = 1

terminos_fibonacci = [f1, f2] 

for i in range(18):
    fn = f1+f2  
    terminos_fibonacci.append(fn)   
    f1 = f2
    f2 = fn  
    
terminos_fibonacci    
    

[1,
 1,
 2,
 3,
 5,
 8,
 13,
 21,
 34,
 55,
 89,
 144,
 233,
 377,
 610,
 987,
 1597,
 2584,
 4181,
 6765]

In [15]:
n = 14
divisores_propios = []

for d in range(1, n): 
    if n%d == 0: 
        divisores_propios.append(d)

if divisores_propios[-1] == 1:
    print('el numero es primo')
else:
    print('el número no es primo')    

el número no es primo



### Ejercicio 2.2 — Media, desviación y outliers con `while` (13 pts)

Usando **solo ciclos `while`**, calcule la **media** y la **desviación estándar** de la lista `datos`. Luego, cuente cuántos valores están fuera del intervalo `[media - 2*desv, media + 2*desv]` (outliers).

⚠️ **No use `sorted()`, `sum()`, `len()`, `min()` ni `max()`** dentro de los cálculos principales (solo `while` puro).



$$\sigma = \sqrt{\frac{1}{N}\sum_{i=1}^{N}(x_i-\mu)^2}$$


In [16]:

# Celda 2.2
datos = [10, 12, 11, 13, 12, 11, 10, 12, 100, 11, 13, 12, 11, 10, 12, 13]

# TU CÓDIGO AQUÍ



In [ ]:
suma = 0


16

In [ ]:

cuenta = 0

for i in datos: 
    cuenta+=1

cuenta

In [22]:
p = 0
suma = 0

while p <= cuenta-1: 
    suma += datos[p]
    p +=1

suma     

273

In [24]:
media = suma/cuenta

media

17.0625

In [25]:
datos

[10, 12, 11, 13, 12, 11, 10, 12, 100, 11, 13, 12, 11, 10, 12, 13]

In [26]:
desviaciones = [datos[i] - media for i in range(cuenta)]
desviaciones 

[-7.0625,
 -5.0625,
 -6.0625,
 -4.0625,
 -5.0625,
 -6.0625,
 -7.0625,
 -5.0625,
 82.9375,
 -6.0625,
 -4.0625,
 -5.0625,
 -6.0625,
 -7.0625,
 -5.0625,
 -4.0625]

In [27]:
cuadrados = [desviaciones[i]**2 for i in range(cuenta)]
cuadrados 

[49.87890625,
 25.62890625,
 36.75390625,
 16.50390625,
 25.62890625,
 36.75390625,
 49.87890625,
 25.62890625,
 6878.62890625,
 36.75390625,
 16.50390625,
 25.62890625,
 36.75390625,
 49.87890625,
 25.62890625,
 16.50390625]

In [28]:
p = 0
suma_de_cuadrados = 0

while p <= cuenta-1: 
    suma_de_cuadrados += cuadrados[p]
    p +=1

suma_de_cuadrados     

7352.9375

In [30]:
import math
desviacion_estandar = math.sqrt(suma_de_cuadrados/cuenta)

desviacion_estandar 

21.43731778348215


## Sección 3: Funciones con `def` (25 pts)

### Ejercicio 3.1 — Factorización en primos y número de divisores (12 pts)

Cree `factorizar(n)` que retorne un **diccionario** con la factorización en primos de `n` (primo → exponente). Luego, cree `num_divisores(n)` que use la factorización para calcular cuántos divisores tiene `n`.

Pruebe con `n = 360`.

```python
# Celda 3.1

# TU CÓDIGO AQUÍ
```


In [32]:
{'nombre':'Marco', 'estatura': 1.65}

{'nombre': 'Marco', 'estatura': 1.65}

In [35]:
n = 8

divisores_propios = []

for d in range(1,n): 
    if n%d==0:
        divisores_propios.append(d)
    
divisores_propios    

[1, 2, 4]

In [36]:
def num_divisores_propios(n):
    divisores_propios = []

    for d in range(1,n): 
        if n%d==0:
            divisores_propios.append(d)
    return len(divisores_propios)        

In [37]:
num_divisores_propios(8)

3

In [40]:
def es_primo(n):
    if num_divisores_propios(n)==1:
        return 'es primo'

In [42]:
es_primo(14)

In [ ]:
factores_primos  = []

for i in divisores_propios:
    


# Sección 4: Funciones `lambda` (15 pts)

### Ejercicio 4.1 — Lambdas para teoría de números (7 pts)

Usando `lambda`, cree funciones para:
1. Determinar si un número es **cuadrado perfecto**: `es_cuadrado`
2. Calcular el **valor absoluto**: `absoluto`
3. Determinar si dos números son **coprimos** (MCD = 1): `son_coprimos`



```python
# Celda 4.1

# TU CÓDIGO AQUÍ
```



### Ejercicio 1.2 — Clasificador de notas con curva (10 pts)

Dado un puntaje `p` (0–100) y una curva configurable `curva` (puntos extra), clasifique la nota final según:
- `>= 90` → `"Excelente"`
- `>= 80` → `"Bueno"`
- `>= 70` → `"Aceptable"`
- `>= 60` → `"Insuficiente"`
- `< 60` → `"Reprobado"`

**Reglas especiales:**
- El puntaje final no puede exceder 100.
- Si el puntaje original es `< 60` pero con la curva alcanza `>= 60`, indique `"Aprobado por curva"`.

**Pruebe con:** `p = 55, curva = 8`.

```python
# Celda 1.2
p = 55
curva = 8

# TU CÓDIGO AQUÍ
```

---



# Sección 5: Integrador — Mini Análisis de Datos (15 pts)

### Ejercicio 5.1 — Análisis de temperaturas semanales (sin `sorted()`)

Dado el siguiente registro de **temperaturas** (en °C) de 4 semanas (7 días cada una):

```python
temperaturas = [
    [22, 24, 26, 25, 23, 21, 20],   # Semana 1
    [19, 21, 23, 24, 22, 20, 18],   # Semana 2
    [25, 27, 28, 29, 27, 26, 24],   # Semana 3
    [18, 17, 19, 21, 20, 19, 22]    # Semana 4
]
```

Realice lo siguiente **usando solo Python estándar y sin `sorted()`**:

1. **Función `clasificar_temp(t)`** con `if/elif/else`:
   - `< 18` → `"Fría"`
   - `18–22` → `"Templada"`
   - `23–27` → `"Cálida"`
   - `>= 28` → `"Calurosa"`

2. **Ciclo `for` anidado** para recorrer la matriz y contar cuántos días hay en cada categoría.

3. **Ciclo `while`** para calcular el promedio general y el promedio de cada semana.

4. **Función `lambda`** para filtrar los días "Calurosos" (`>= 28`).

5. **Función `min_max_sin_sorted(datos)`** que retorne el mínimo y máximo de una lista plana **sin usar `sorted()` ni `min()`/`max()`** (solo `for` y condicionales). Aplíquela a todos los datos aplanados.

6. Imprimir un **reporte** consolidado.

```python
# Celda 5.1
temperaturas = [
    [22, 24, 26, 25, 23, 21, 20],
    [19, 21, 23, 24, 22, 20, 18],
    [25, 27, 28, 29, 27, 26, 24],
    [18, 17, 19, 21, 20, 19, 22]
]

# TU CÓDIGO AQUÍ
```



# Distribución del Tiempo Sugerida

| Sección | Tiempo |
|---------|--------|
| 1. Condicionales | 15 min |
| 2. Ciclos | 20 min |
| 3. Funciones `def` | 25 min |
| 4. Lambdas | 15 min |
| 5. Integrador | 25 min |

---
